geoips/pydantic/bases.py

```diff
@@ -0,0 +1,156 @@
+"""Implementations of Pydantic base models for GeoIPS.
+
+The pydantic base models defined here are intended for use by other base
+models.
+
+``Plugin`` should be used as the parent class of all other plugin models.
+
+The other models defined here are intended to validate particular field types
+within plugin models.
+"""
+
+# Python Standard Libraries
+import keyword
+from pathlib import Path
+from typing import Tuple
+
+
+# Third-Party Libraries
+from matplotlib.artist import Artist
+from matplotlib.lines import Line2D
+from pydantic import BaseModel, ConfigDict, Field, field_validator
+from pydantic_core import PydanticCustomError
+from pydantic.functional_validators import AfterValidator
+from typing_extensions import Annotated
+
+
+class PrettyBaseModel(BaseModel):
+    """Make Pydantic models pretty-print by default."""
+
+    def __str__(self) -> str:
+        """Return a string representation of a Pydantic model.
+
+        The returned string will be formatted as JSON with two-space indentation.
+
+        Returns
+        -------
+            str: A string representation of the Pydantic model.
+        """
+        return self.model_dump_json(indent=2)
+
+
+def python_identifier(val: str) -> str:
+    """Validate if a string is a valid Python identifier.
+
+    Validate if a string is a valid Python identifier and not a reserved Python keyword.
+    See https://docs.python.org/3/reference/lexical_analysis.html#identifiers for more
+    information on Python identifiers and reserved keywords.
+
+    Validation is performed by calling `str.isidentifier` and `keyword.iskeyword`.
+
+    Parameters
+    ----------
+        val: str
+            The input string to validate.
+
+    Returns
+    -------
+        str: The input string if it is a valid Python identifier.
+
+    Raises
+    ------
+        ValueError: If the input string is not a valid Python identifier.
+    """
+    if not val.isidentifier():
+        raise ValueError(f"{val} is not a valid Python identifier")
+    if keyword.iskeyword(val):
+        raise ValueError(f"{val} is a reserved Python keyword")
+    return val
+
+
+# Create the PythonIdentifier type
+PythonIdentifier = Annotated[str, AfterValidator(python_identifier)]
+
+
+class PluginModel(PrettyBaseModel):
+    """Base Plugin model for all GeoIPS plugins."""
+
+    model_config = ConfigDict(extra="forbid")
+    interface: PythonIdentifier = Field(
+        ..., description="The name of the plugin's interface."
+    )
+    family: PythonIdentifier = Field(..., description="The family of the plugin.")
+    name: PythonIdentifier = Field(..., description="The name of the plugin.")
+    # Should write a test to ensure this is a valid numpy docstring
+    docstring: str = Field(..., description="The docstring for the plugin.")
+    package: PythonIdentifier = Field(
+        None, description="The package the plugin belongs to."
```

Comment:
For optional set-by-validators, do we want to error on variables if set by the user?

Reply:
* ditto for rel/abspath and type